File: core/src/main/java/com/xmartlabs/base/core/module/GsonModule.java
package com.xmartlabs.base.core.module;

import android.support.annotation.NonNull;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.xmartlabs.base.core.model.BuildInfo;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class GsonModule {
  @Provides
  @Singleton
  public Gson provideGson(GsonBuilder gsonBuilder) {
    return gsonBuilder.create();
  }

  @NonNull
  @Provides
  public final GsonBuilder provideCommonGsonBuilder(@NonNull BuildInfo buildInfo) {
    return modifyGsonBuilder(new GsonBuilder(), buildInfo);
  }

  protected GsonBuilder modifyGsonBuilder(GsonBuilder builder, @NonNull BuildInfo buildInfo) {
    if (buildInfo.isDebug()) {
      builder.setPrettyPrinting();
    }
    return builder;
  }
}

File: core/src/main/java/com/xmartlabs/bigbang/core/module/GsonModule.kt
package com.xmartlabs.bigbang.core.module

import com.google.gson.GsonBuilder
import com.xmartlabs.bigbang.core.common.GsonExcludeStrategy
import com.xmartlabs.bigbang.core.model.BuildInfo
import dagger.Module
import dagger.Provides
import java.util.*
import javax.inject.Singleton

@Module
open class GsonModule {
  @Provides
  @Singleton
  open fun provideGson(gsonBuilder: GsonBuilder) = gsonBuilder
      .setExclusionStrategies(getExclusionStrategy(null))
      .create()

  @Provides
  open fun provideCommonGsonBuilder(buildInfo: BuildInfo) = modifyGsonBuilder(GsonBuilder(), buildInfo)

  protected open fun modifyGsonBuilder(builder: GsonBuilder, buildInfo: BuildInfo) =
    if (buildInfo.isDebug) builder.setPrettyPrinting() else builder

  protected open fun getExclusionStrategy(strategy: GsonExcludeStrategy?) =
      GsonExclusionStrategy(ArrayList()).getExclusionStrategy(strategy)
}
